File: 프로그래머스/3/64062. 징검다리 건너기/징검다리 건너기.java
import java.util.*;
class Solution {
      public int solution(int[] stones, int k) {
            // 이분탐색으로 해야 효율성을 클리어 할 수 있다고 한다...
            // 최소 최대를 이용해 중간값 구하기..
            int min = Integer.MAX_VALUE;
            int max = Integer.MIN_VALUE;
            for (int stone : stones) {
                min = Math.min(min, stone);
                max = Math.max(max, stone);
            }
            while(min < max){
                int mid = (min + max + 1) / 2;
                if(walk(mid, k, stones)){
                    min = mid;
                }else{
                    max = mid - 1;
                }
            }

            return max;
        }
        public boolean walk(int mid, int k, int[] stones){
            int cnt = 0;
            for(int stone : stones){
                if(stone - mid < 0){
                    cnt++;
                }else{
                    cnt = 0;
                }

                if(cnt == k){
                    return false;
                }
            }
            return true;
        }
}

File: 프로그래머스/3/64062. 징검다리 건너기/징검다리 건너기.py
def solution(stones, k):
    answer = 0
    min_stone = min(stones)
    max_stone = max(stones)
    def walk(mid, cnt, stones):
        for stone in stones:
            if stone < mid:
                cnt += 1
            else:
                cnt = 0
            if cnt == k:
                return False
        return True
    while min_stone < max_stone:
        mid_stone = (min_stone + max_stone + 1) // 2
        check = walk(mid_stone, 0, stones)
        # 중간값이 실제 stone들보다 큰 거니깐 min을 늘리는 방식인건가..?
        if check:
            min_stone = mid_stone
        # 중간값이 실제 stone들보다 작은거니깐 max를 늘리는 형태고
        else:
            max_stone = mid_stone - 1
    return max_stone